tests/builder/subcommands.rs

```diff
@@ -635,3 +635,184 @@ fn duplicate_subcommand_alias() {
         .subcommand(Command::new("unique").alias("repeat"))
         .build();
 }
+
+#[test]
+fn test_help_header() {
+    static VISIBLE_ALIAS_HELP: &str = "\
+Usage: clap-test [COMMAND]
+
+Commands:
+  help  Print this message or the help of the given subcommand(s)
+
+Test commands:
+  test  Some help
+
+Options:
+  -h, --help     Print help
+  -V, --version  Print version
+";
+    
+    let cmd = Command::new("clap-test")
+    .version("2.6")
+    .subcommand(
+        Command::new("test")
+        .about("Some help")
+        .help_heading("Test commands"),
+    );
+    utils::assert_output(cmd, "clap-test --help", VISIBLE_ALIAS_HELP, false);
+}
+
+#[test]
+fn test_help_header_multiple_help_headers() {
+    static VISIBLE_ALIAS_HELP: &str = "\
+Usage: clap-test [COMMAND]
+
+Commands:
+  help   Print this message or the help of the given subcommand(s)
+
+Test Commands 1:
+  test1  Some help
+
+Test Commands 2:
+  test2  Some help
+
+Options:
+  -h, --help     Print help
+  -V, --version  Print version
+";
+
+    let cmd = Command::new("clap-test")
+        .version("2.6")
+        .subcommand(
+            Command::new("test1")
+                .about("Some help")
+                .help_heading("Test Commands 1"),
+        )
+        .subcommand(
+            Command::new("test2")
+                .about("Some help")
+                .help_heading("Test Commands 2"),
+        );
+
+    utils::assert_output(cmd, "clap-test --help", VISIBLE_ALIAS_HELP, false);
+}
+
+#[test]
+fn test_help_header_hide_commands_header() {
+    static VISIBLE_ALIAS_HELP: &str = "\
+Usage: clap-test [COMMAND]
+
+Test commands:
+  test  Some help
+
+Options:
+  -h, --help     Print help
+  -V, --version  Print version
+";
+
+    let cmd = Command::new("clap-test")
+        .version("2.6")
+        .disable_help_subcommand(true)
+        .subcommand_help_heading("Test commands")
+        .subcommand(Command::new("test").about("Some help"));
+    utils::assert_output(cmd, "clap-test --help", VISIBLE_ALIAS_HELP, false);
+}
+
+#[test]
+fn test_multiple_commands_mixed_standard_and_custom_headers() {
```

Comment:
We don't have a test that defines commands with help headings out of order (e.g. a `First Custom`  following by a `Second Custom` followed by a `First Custom`).

Reply:
Shuffled the order of the user commands in the comprehensive mixed test.